File: src/main/java/data/analysis/Stationary.java
package data.analysis;

import data.Line;

import java.util.ArrayList;

public class Stationary {
    public static double avg(final Line line, int startIdx, int endIdx){
        var result = 0.0;
        for (int i = startIdx; i < endIdx; i++) {
            result += line.getY(i);
        }
        return result/(endIdx-startIdx);
    }
    public static double dispersion(final Line line, int startIdx, int endIdx){
        var result = 0.0;
        var avg = avg(line,startIdx,endIdx);
        for (int i = startIdx; i < endIdx; i++) {
            var value = line.getY(i);
            result += (value-avg)*(value-avg);
        }
        return result/(endIdx-startIdx);
    }
    public static double amplitude(final Line line, int startIdx, int endIdx){
        double minVal = line.getY(startIdx);
        double maxVal = line.getY(startIdx);
        for (int i = startIdx + 1; i < endIdx; i++) {
            double curr = line.getY(i);
            if (curr > maxVal) maxVal = curr;
            if (curr < minVal) minVal = curr;
        }
        return maxVal - minVal;
    }
    public static ArrayList<Line> dataPerInterval(final Line line, int intervalsCount, double error){
        Line dispResult = new Line(intervalsCount);
        Line avgResult = new Line(intervalsCount);
        Line errorResult = new Line(intervalsCount);
        final var size = line.getSize();
        final int pointsInInterval = size / intervalsCount;

        int startIdx = 0, endIdx;
        double ampl = amplitude(line,0,size);
        for (int i = 0; i < intervalsCount; i++) {
            endIdx = Math.min(startIdx+pointsInInterval,size);
            dispResult.setX(i,dispersion(line,startIdx,endIdx));
            avgResult.setX(i,avg(line,startIdx,endIdx));
            errorResult.setX(i, error*ampl);
            startIdx+=pointsInInterval;
        }
        var list = new ArrayList<Line>();
        list.add(dispResult);
        list.add(avgResult);
        list.add(errorResult);

        return list;
    }
}

File: src/main/java/data/analysis/Stationary.kt
package data.analysis

import data.Line
import kotlin.math.min

object Stationary {
    fun avg(line: Line, startIdx: Int, endIdx: Int): Double {
        var result = 0.0
        for (i in startIdx until endIdx) {
            result += line.getY(i)
        }
        return result / (endIdx - startIdx)
    }

    fun dispersion(line: Line, startIdx: Int, endIdx: Int): Double {
        var result = 0.0
        val avg = avg(line, startIdx, endIdx)
        for (i in startIdx until endIdx) {
            val value = line.getY(i)
            result += (value - avg) * (value - avg)
        }
        return result / (endIdx - startIdx)
    }

    fun amplitude(line: Line, startIdx: Int, endIdx: Int): Double {
        var minVal = line.getY(startIdx)
        var maxVal = line.getY(startIdx)
        for (i in startIdx + 1 until endIdx) {
            val curr = line.getY(i)
            if (curr > maxVal) maxVal = curr
            if (curr < minVal) minVal = curr
        }
        return maxVal - minVal
    }

    fun dataPerInterval(line: Line, intervalsCount: Int, error: Double): List<Line> {
        val dispResult = Line(intervalsCount)
        val avgResult = Line(intervalsCount)
        val errorResult = Line(intervalsCount)
        val size = line.size
        val pointsInInterval = size / intervalsCount

        var startIdx = 0
        var endIdx: Int
        val ampl = amplitude(line, 0, size)
        for (i in 0 until intervalsCount) {
            endIdx = min(startIdx + pointsInInterval, size)
            dispResult.setX(i, dispersion(line, startIdx, endIdx))
            avgResult.setX(i, avg(line, startIdx, endIdx))
            errorResult.setX(i, error * ampl)
            startIdx += pointsInInterval
        }
        return listOf(dispResult, avgResult,errorResult)
    }
}
